src/extendedstorage.cpp

```diff
@@ -417,184 +414,156 @@ void ExtendedStorage::setUpdated(const KCalendarCore::Incidence::List &added,
+bool ExtendedStorage::addNotebook(const Notebook &nb)
 {
-    if (nb->uid().length() < 7) {
-        // Cannot accept this id, create better one.
-        QString uid(QUuid::createUuid().toString());
-        nb->setUid(uid.mid(1, uid.length() - 2));
-    }
-
-    if (!nb || d->mNotebooks.contains(nb->uid())) {
+    if (!nb.isValid() || d->mNotebooks.contains(nb.uid())) {
         return false;
     }
 
-    if (!modifyNotebook(nb, DBInsert)) {
+    if (!nb.isRunTimeOnly() && !modifyNotebook(nb, DBInsert)) {
         return false;
     }
+    d->mNotebooks.insert(nb.uid(), nb);
 
-    d->mNotebooks.insert(nb->uid(), nb);
-    if (!calendar()->addNotebook(nb->uid(), nb->isVisible())
-        && !calendar()->updateNotebook(nb->uid(), nb->isVisible())) {
-        qCWarning(lcMkcal) << "notebook" << nb->uid() << "already in calendar";
+    if (!calendar()->addNotebook(nb.uid(), nb.isVisible())
+        && !calendar()->updateNotebook(nb.uid(), nb.isVisible())) {
+        qCWarning(lcMkcal) << "notebook" << nb.uid() << "already in calendar";
     }
 
     return true;
 }
 
-bool ExtendedStorage::updateNotebook(const Notebook::Ptr &nb)
+bool ExtendedStorage::updateNotebook(const Notebook &nb)
 {
-    if (!nb || !d->mNotebooks.contains(nb->uid()) ||
-            d->mNotebooks.value(nb->uid()) != nb) {
+    if (!nb.isValid() || !d->mNotebooks.contains(nb.uid())) {
         return false;
     }
 
-    bool wasVisible = calendar()->isVisible(nb->uid());
-    if (!calendar()->updateNotebook(nb->uid(), nb->isVisible())) {
-        qCWarning(lcMkcal) << "cannot update notebook" << nb->uid() << "in calendar";
+    if (!nb.isRunTimeOnly() && !modifyNotebook(nb, DBUpdate)) {
         return false;
     }
-    if (!modifyNotebook(nb, DBUpdate)) {
-        return false;
+    d->mNotebooks.insert(nb.uid(), nb);
+
+    bool wasVisible = calendar()->isVisible(nb.uid());
+    if (!calendar()->updateNotebook(nb.uid(), nb.isVisible())) {
+        qCWarning(lcMkcal) << "cannot update notebook" << nb.uid() << "in calendar";
     }
 
 #if defined(TIMED_SUPPORT)
-    if (wasVisible && !nb->isVisible()) {
-        d->clearAlarms(nb->uid());
-    } else if (!wasVisible && nb->isVisible()) {
-        Incidence::List list;
-        if (allIncidences(&list, nb->uid())) {
-            d->setAlarmsForNotebook(list, nb->uid());
+    if (!nb.isRunTimeOnly()) {
+        if (wasVisible && !nb.isVisible()) {
+            d->clearAlarms(nb.uid());
+        } else if (!wasVisible && nb.isVisible()) {
+            Incidence::List list;
+            if (allIncidences(&list, nb.uid())) {
+                d->setAlarmsForNotebook(list, nb.uid());
+            }
         }
     }
 #endif
 
     return true;
 }
 
-bool ExtendedStorage::deleteNotebook(const Notebook::Ptr &nb)
+bool ExtendedStorage::deleteNotebook(const QString &nbid)
 {
-    if (!nb || !d->mNotebooks.contains(nb->uid())) {
-        return false;
+    if (!d->mNotebooks.contains(nbid)) {
+        return true;
```

Comment:
Just a small note to check that this change in semantics (i.e. it now returns `true` if the notebook doesn't exist, rather than `false`) is intentional. This seems to potentially have implications for EAS (which can be worked around).

Reply:
Yes, it's intentional. If the notebook doesn't exist, then the intent of not having it anymore in the DB is a success.

If it's too much of a pain for EAS, one can still go back to returning `false`. I'm not opposed to it.

Reply:
Oh, by the way, there is now a `bool ExtentedStorage::containsNotebook(const QString &nbuid)` so you can get the exact same behaviour as before in EAS by replacing:

- `if (!deleteNotebook(nb))`, with
- `if (!containsNotebook(nb.uid()) || !deleteNotebook(nb.uid()))`

Reply:
Thanks, I'd missed `containsNotebook`. That does make things nicer elsewhere, yes.

Reply:
Yes, besides the basic changes from pointers to references, the changes are :
- `if (notebook)` becomes `if (notebook.isValid())`
- `deleteNotebook(nb)` becomes `deleteNotebook(nb.uid())`
- `Notebook::Ptr notebook = storage->notebook(uid); if (notebook)` becomes `if (containsNotebook(uid))` in the case `notebook` is not used thereafter
- `Notebook::Ptr notebook = storage->notebook(uid); ...; storage->update(notebook);` this construction fetching the notebook before updating it, is not necessary anymore. The `notebook` argument passed to the storage is not a pointer anymore to the actual notebook object used in the storage. The update is done by copy based on the uid.